File: riak_client/verify_graph_coloring.py
import time

from custom_logger import logger
from graph_helpers import get_graph
from graph_coloring import get_args_parser
from client_helpers import get_partition_for_client
from riak_helpers import (
    RIAK_BUCKET_PREFIX,
    RIAK_NODE_KEY_PREFIX,
    get_request_riak,
)


def verify(graph, node):
    self_color = get_request_riak(
        RIAK_BUCKET_NAME,
        f"{RIAK_NODE_KEY_PREFIX}{node}__val",
    )
    if not (0 <= self_color <= graph.degree(node)):  # strict checks
        raise Exception(f"Invalid color {self_color} assigned to the node {node}.")

    for nbr in graph.neighbors(node):
        nbr_color = get_request_riak(
            RIAK_BUCKET_NAME,
            f"{RIAK_NODE_KEY_PREFIX}{nbr}__val",
        )
        if not (
            0 <= nbr_color <= graph.degree(nbr)
        ):  # strict checks to avoid none values
            raise Exception(
                f"Invalid color {nbr_color} assigned to the neighbor {nbr} of node {node}."
            )

        if self_color == nbr_color:
            return False

    return True


def main(graph):
    failed_count = 0
    passed_count = 0

    for node in CLIENT_NODES:
        passed = verify(graph, node)
        if passed:
            passed_count += 1
            logger.debug(f"Node {node} passed verification.")
        else:
            failed_count += 1
            logger.warning(f"Node {node} failed verification.")

    logger.info(
        f"Verification complete: Total : {len(CLIENT_NODES)} | Passed: {passed_count} | Failed: {failed_count}."
    )


if __name__ == "__main__":
    start_time = time.time()
    args = get_args_parser()
    graph_name = args.graph_name
    graph = get_graph(graph_name)
    logger.info(f"Found graph {graph}.")
    if args.client_id >= args.num_clients or args.client_id < 0:
        raise Exception("Client ID must be in the range [0, num_clients-1].")
    client_id = args.client_id
    num_clients = args.num_clients
    CLIENT_NODES = get_partition_for_client(graph, client_id, num_clients)
    logger.info(f"Client {client_id} handling nodes: {CLIENT_NODES}.")
    RIAK_BUCKET_NAME = f"{RIAK_BUCKET_PREFIX}__{graph_name}"
    logger.info(f"Using Riak bucket: {RIAK_BUCKET_NAME}")
    main(graph)
    logger.info(f"Total time taken: {time.time() - start_time} seconds.")

File: riak_client/erlang_imp/src/verify_graph_coloring.erl
-module('verify_graph_coloring').
-export([start/0]).
-include("graph.hrl").

start() ->
    my_logger:setup(),
    Args = init:get_arguments(),
    case
        {
            proplists:get_value('graph-name', Args),
            proplists:get_value('client-id', Args),
            proplists:get_value('num-clients', Args)
        }
    of
        {undefined, _, _} ->
            usage();
        {_, undefined, _} ->
            usage();
        {_, _, undefined} ->
            usage();
        {GraphName, ClientIdStr, NumClientsStr} ->
            {ok, ClientId} = graph_coloring:safe_int(ClientIdStr),
            {ok, NumClients} = graph_coloring:safe_int(NumClientsStr),
            my_logger:info(
                io_lib:format(
                    "Starting graph coloring with graph: ~s, client ID: ~p, number of clients: ~p",
                    [GraphName, ClientId, NumClients]
                )
            ),
            Graph = graph:get_graph(GraphName),
            my_logger:info(
                io_lib:format("Loaded graph: ~s", [graph:to_string(Graph)])
            ),
            Partition = partition:get_partition_for_client(
                graph:nodes(Graph), ClientId, NumClients
            ),
            my_logger:info(
                io_lib:format("Client ~p has partition: ~p", [ClientId, Partition])
            ),
            main(Graph, Partition)
    end.

usage() ->
    my_logger:warning(
        io_lib:format(
            "Usage: erl -noshell -s verify_graph_coloring start -graph-name <name> -client-id <id> -num-clients <num> -s init stop",
            []
        )
    ),
    halt(1).

verify([], _, _) ->
    true;
verify([Nbr | Rest], SelfColor, Graph) ->
    NbrColor = riak_client:get_request_riak(
        io_lib:format("graph_coloring__~s", [Graph#graph.name]),
        io_lib:format("node_~p__val", [Nbr])
    ),
    case NbrColor of
        undefined ->
            my_logger:warning(
                io_lib:format("Neighbor node ~p has no color assigned!", [Nbr])
            ),
            false;
        SelfColor ->
            false;
        _ ->
            verify(Rest, SelfColor, Graph)
    end.

verifyNode([], _, Acc) ->
    Acc;
verifyNode([Node | Rest], Graph, {Passed, Failed}) ->
    Color = riak_client:get_request_riak(
        io_lib:format("graph_coloring__~s", [Graph#graph.name]),
        io_lib:format("node_~p__val", [Node])
    ),
    case Color of
        undefined ->
            my_logger:warning(
                io_lib:format("Node ~p has no color assigned!", [Node])
            ),
            verifyNode(Rest, Graph, {Passed, Failed + 1});
        _ ->
            Neighbors = graph_coloring:get_lexically_ordered_neighbors(Graph, Node),
            case verify(Neighbors, Color, Graph) of
                true ->
                    verifyNode(Rest, Graph, {Passed + 1, Failed});
                false ->
                    my_logger:warning(
                        io_lib:format("Coloring conflict found at node ~p!", [Node])
                    ),
                    verifyNode(Rest, Graph, {Passed, Failed + 1})
            end
    end.

main(Graph, Partition) ->
    {PassedCount, FailedCount} = verifyNode(Partition, Graph, {0, 0}),
    io:format("Passed: ~p, Failed: ~p~n", [PassedCount, FailedCount]),
    my_logger:info(
        io_lib:format(
            "Verification complete. Passed: ~p, Failed: ~p",
            [PassedCount, FailedCount]
        )
    ),
    ok.
